[PROCS-3728] Modify process check to populate process with the process context

levin0710 opened the pull request:
<!--
* New contributors are highly encouraged to read our
  [CONTRIBUTING](/CONTRIBUTING.md) documentation.
* The pull request:
  * Should only fix one issue or add one feature at a time.
  * Must update the test suite for the relevant functionality.
  * Should pass all status checks before being reviewed or merged.
* Commit titles should be prefixed with general area of pull request's change.

-->
### What does this PR do?

This PR modifies the process check to populate process with the newly created field "ProcessContext", which has been added to agent-payload in the following [PR](https://github.com/DataDog/agent-payload/pull/284).

<!--
* A brief description of the change being made with this pull request.
* If the description here cannot be expressed in a succinct form, consider
  opening multiple pull requests instead of a single one.
-->

### Motivation
Process level data such as Process-context needs to be integrated into the process check, as outlined in the following JIRA ticket: https://datadoghq.atlassian.net/browse/PROCS-3728
<!--
* What inspired you to submit this pull request?
* Link any related GitHub issues or PRs here.
-->

### Additional Notes

<!--
* Anything else we should know when reviewing?
* Include benchmarking information here whenever possible.
* Include info about alternatives that were considered and why the proposed
  version was chosen.
-->



### Possible Drawbacks / Trade-offs

<!--
* What are the possible side-effects or negative impacts of the code change?
-->

### Describe how to test/QA your changes

<!--
* Write here in detail or link to detailed instructions on how this change can
  be tested/QAd/validated, including any environment setup.
-->

Using the standard config, run the following steps:

- Run the Process check
- Verify that processes contain the `ProcessContext` attribute
- Verify that `ProcessContext` looks something like this: `ProcessContext: ["process_context:sh"]`

### Reviewer's Checklist
<!--
* Authors can use this list as a reference to ensure that there are no problems
  during the review but the signing off is to be done by the reviewer(s).

Note: Adding GitHub labels is only possible for contributors with write access.
-->

- [x] If known, an appropriate milestone has been selected; otherwise the `Triage` milestone is set.
- [ ] Use the `major_change` label if your change either has a major impact on the code base, is impacting multiple teams or is changing important well-established internals of the Agent. This label will be use during QA to make sure each team pay extra attention to the changed behavior. For any customer facing change use a releasenote.
- [x] A [release note](https://github.com/DataDog/datadog-agent/blob/main/docs/dev/contributing.md#reno) has been added or the `changelog/no-changelog` label has been applied.
- [x] Changed code has automated tests for its functionality.
- [x] Adequate QA/testing plan information is provided. Except if the `qa/skip-qa` label, with required either `qa/done` or `qa/no-code-change` labels, are applied.
- [x] At least one `team/..` label has been applied, indicating the team(s) that should QA this change.
- [ ] If applicable, docs team has been notified or [an issue has been opened on the documentation repo](https://github.com/DataDog/documentation/issues/new).
- [ ] If applicable, the `need-change/operator` and `need-change/helm` labels have been applied.
- [ ] If applicable, the `k8s/<min-version>` label, indicating the lowest Kubernetes version compatible with this feature.
- [ ] If applicable, the [config template](https://github.com/DataDog/datadog-agent/blob/main/pkg/config/config_template.yaml) has been updated.

levin0710 committed:
implemented change

levin0710 committed:
add test

levin0710 committed:
fix functionality and tests

levin0710 committed:
Merge branch 'main' into levin.sanchez/payload-experimentation

levin0710 committed:
enable extractor by default

levin0710 committed:
remove unused config

levin0710 committed:
remove unnecessary variables

levin0710 committed:
rename symbols

levin0710 committed:
modify to use toggle for service estractor

levin0710 committed:
refractor service extractor config and fixed tests

levin0710 committed:
Merge branch 'main' into levin.sanchez/payload-experimentation

levin0710 committed:
Merge branch 'main' into levin.sanchez/payload-experimentation

levin0710 committed:
update agent version

levin0710 committed:
reno

levin0710 committed:
format

levin0710 committed:
format

levin0710 committed:
fix windows tests

levin0710 committed:
Delete my-server.sh

levin0710 committed:
fix windows test

levin0710 committed:
Merge branch 'levin.sanchez/payload-experimentation' of github.com:DataDog/datadog-agent into levin.sanchez/payload-experimentation

daniel-taf commented on pkg/process/checks/net.go:
```diff
@@ -99,7 +99,11 @@ func (c *ConnectionsCheck) Init(syscfg *SysProbeConfig, hostInfo *HostInfo, _ bo
 	c.processData = NewProcessData(c.config)
 	c.dockerFilter = parser.NewDockerProxy()
-	c.serviceExtractor = parser.NewServiceExtractor(c.sysprobeYamlConfig)
+	var (
+		enabled               = c.sysprobeYamlConfig.GetBool("system_probe_config.process_service_inference.enabled")
+		useWindowsServiceName = c.sysprobeYamlConfig.GetBool("system_probe_config.process_service_inference.use_windows_service_name")
+	)
+	c.serviceExtractor = parser.NewServiceExtractor(enabled, useWindowsServiceName)
```
No need to use var here or in other places. You can use `:=` instead. For example:
```
enabled := c.sysprobeYamlConfig.GetBool("system_probe_config.process_service_inference.enabled")
```

wiyu replied:
Why are we changing this? Passing the system probe config allowed the service extractor to figure out what configs it needed to be initialized and encapsulated the behavior.

levin0710 replied:
We wanted to be able to set the values of the ServiceExtractor arbitrarly and not be tied to a specific type of config. This is because the process check did not have access to the system-probe config. We discussed it in this thread @wiyu https://dd.slack.com/archives/C0201FP7SPM/p1707423196154039

wiyu replied:
This would be cleaner in a function to encapsulate the behavior.

daniel-taf commented on pkg/process/checks/process.go:
```diff
@@ -147,6 +150,9 @@ func (p *ProcessCheck) Init(syscfg *SysProbeConfig, info *HostInfo, oneShot bool
 
 	p.initConnRates()
 
+	p.serviceExtractor = parser.NewServiceExtractor(true, true)
```
I think its a bit cleaner here to use variables instead of literal booleans. Makes it more clear what you trying to do.

levin0710 replied:
updated!

daniel-taf commented on pkg/process/checks/process_test.go:
```diff
@@ -117,6 +123,12 @@ func TestProcessCheckSecondRun(t *testing.T) {
 	proc3 := makeProcessWithCreateTime(3, "foo --version", now+2)
 	proc4 := makeProcessWithCreateTime(4, "foo -bar -bim", now+3)
 	proc5 := makeProcessWithCreateTime(5, "datadog-process-agent --cfgpath datadog.conf", now+2)
+
+	proc1.ProcessContext = []string{"process_context:git"}
+	proc2.ProcessContext = []string{"process_context:mine-bitcoins"}
+	proc3.ProcessContext = []string{"process_context:foo"}
+	proc4.ProcessContext = []string{"process_context:foo"}
+	proc5.ProcessContext = []string{"process_context:datadog-process-agent"}
```
Right now you have two different approaches (`makeProcessWithProcessContext` or just assigning the field) for adding the process context to procs. I think you should choose one way for consistency. I think what makes most sense is updating `makeProcessWithCreateTime` to accept a context parameter and set it as it seems its always needed.

levin0710 replied:
Decided to not do either and go with an approach where procutil does not have to be modified

robertjli commented on pkg/process/procutil/process_model.go:
```diff
@@ -13,19 +13,19 @@ import (
+	Username       string // (Windows only)
+	Uids           []int32
+	Gids           []int32
+	ProcessContext []string
```
IMO we shouldn't add this to the procutil model. This data is not collected by the procutil package, and this field is not used in the actual agent, only in the tests that you added (which can be done another way).

levin0710 replied:
Modified test to not rely on procutil modification

robertjli commented on pkg/process/checks/process_test.go:
```diff
@@ -541,3 +554,32 @@ func TestProcessCheckZombieToggleTrue(t *testing.T) {
+	actual, err := processCheck.run(0, false)
+	require.NoError(t, err)
+	assert.ElementsMatch(t, expected, actual.Payloads())
+
```
Nit: remove extra newline

levin0710 replied:
updated

levin0710 committed:
addressing PR comments

levin0710 committed:
nits

levin0710 commented on pkg/process/metadata/parser/service_test.go:
```diff
@@ -130,8 +130,11 @@ func TestExtractServiceMetadata(t *testing.T) {
-			se := NewServiceExtractor(mockConfig)
+			var (
+				enabled               = mockConfig.GetBool("system_probe_config.process_service_inference.enabled")
+				useWindowsServiceName = mockConfig.GetBool("system_probe_config.process_service_inference.use_windows_service_name")
```
updated!

richl-dd commented on pkg/process/metadata/parser/service_windows_test.go:
💭 / ❓: I'm not sure exactly if this is within scope for this PR, however, since it modifies the `_test.go` file, I figured I'd ask: do you know if the tests here are actually verifying that for a given pid, we're exercising the mechanisms that look up the service name as reflected by the SCM?

@just-chillin, looking back, it seems like you were the original author of the Windows SCM PR (https://github.com/DataDog/datadog-agent/pull/15615), from what I can tell, we may not be using the actual SCM service extractor and instead using a mocked one. If we're not using the actual extractor, then I think it might be worth adding a test that does if we're planning on adding the `ProcessContext` to the process payload and enabling `use_windows_service_name`  by default. If the actual service extractor is tested somewhere else, then this is probably less important, but I couldn't identify where it might be being tested.

If we're not (and we want to), an example for how to add this to the tests could look something like executing a command via PowerShell to list all the Windows service names and their pid in JSON and then comparing the command output to what we report in the actual extractor (again, if we're doing this somewhere else and this is a case of my not being able to find it, we can ignore the following 😅):

```powershell
Get-WmiObject Win32_Service | Select-Object Name, ProcessId | ConvertTo-Json
```

which outputs something like:
```json
[
    {
        "Name":  "MyServiceFoo",
        "ProcessId":  1234
    },
    {
        "Name":  "MyServiceBar",
        "ProcessId":  4567
    }
]
```

From that, we could pass a couple of the returned `ProcessId`s to service extractor and verify that for a given `ProcessId`, we report the same service name as shown in the command.

levin0710 replied:
I think we should add that test. I was wondering how would you get `Get-WmiObject Win32_Service | Select-Object Name, ProcessId | ConvertTo-Json` in code for the test?

richl-dd replied:
So, your tests are running inside a Windows container, which means you can execute commands from Go in the same way that you would if you were on a Linux host / container or anywhere else via the `os/exec` package. To be fair, this blurs the lines a _little_ between unit / integration tests, however, in this scenario I think its alright (the alternative would be to try and write an E2E test, which for this PR, would be a lot, but might be worth adding in the future). Anyway, it'd look something like this in your test code:

```go
type Service struct {
    Name      string `json:"Name"`
    ProcessId int    `json:"ProcessId"`
}

// PowerShell command to execute
psCmd := `Get-WmiObject Win32_Service | Select-Object Name, ProcessId | ConvertTo-Json`

// Run the command
cmd := exec.Command("powershell", "-Command", psCmd)
var out bytes.Buffer
cmd.Stdout = &out
err := cmd.Run()
if err != nil {
	fmt.Println("Error executing command:", err)
	return
}

// Deserialize JSON output into slice of Service structs
var services []Service
err = json.Unmarshal(out.Bytes(), &services)
if err != nil {
	fmt.Println("Error parsing JSON:", err)
	return
}

// Print the services
fmt.Println("Services:")
for _, service := range services {
	fmt.Printf("Name: %s, ProcessId: %d\n", service.Name, service.ProcessId)
}
```

From there, you'll have all the services that are running inside the Windows container where your test is running. In an ideal world, all the services would stay around between the time you poll them and compare with what our extractor sees, however, there is a chance that some would start/stop in-between, which means they'd have different pids and the test would be flaky. So, I'd suggest choosing services that must be running and not stoppable. Two that come to mind are `DcomLaunch` and `Eventlog` (those are the service names). That should get you going - let me know if you run into any issues!

richl-dd replied:
Understood + agree - adding tests / verifying functionality makes sense to be its own PR 😄

daniel-taf commented on pkg/process/checks/net_test.go:
```diff
@@ -55,7 +55,10 @@ func TestDNSNameEncoding(t *testing.T) {
 		"1.1.2.5": {Names: nil},
 	}
-	ex := parser.NewServiceExtractor(ddconfig.MockSystemProbe(t))
+	conf := ddconfig.MockSystemProbe(t)
+	enabled := conf.GetBool("system_probe_config.process_service_inference.enabled")
+	useWindowsServiceName := conf.GetBool("system_probe_config.process_service_inference.use_windows_service_name")
+	ex := parser.NewServiceExtractor(enabled, useWindowsServiceName)
```
Just like in `service_test.go`, you don't need to use a mock here or in any other test using the service extractor. The constructor just accepts boolean values which you can use directly.

levin0710 replied:
specifically in `net.go`, arent some of the tests also testing if it correctly sources the value from the system-probe? Like the one in line 627?

daniel-taf replied:
To clarify, you shouldn't need to use a mock config **specifically** for getting the service extractor values anymore. If the mock config is used for other purposes in the test you should keep it.

levin0710 replied:
ok, think i addressed this now

levin0710 committed:
renamed variables and removed unnecessary use of mock configs

levin0710 commented on pkg/process/checks/net.go:
```diff
@@ -99,7 +99,9 @@ func (c *ConnectionsCheck) Init(syscfg *SysProbeConfig, hostInfo *HostInfo, _ bo
 	c.dockerFilter = parser.NewDockerProxy()
-	c.serviceExtractor = parser.NewServiceExtractor(c.sysprobeYamlConfig)
+	enabled := c.sysprobeYamlConfig.GetBool("system_probe_config.process_service_inference.enabled")
+	useWindowsServiceName := c.sysprobeYamlConfig.GetBool("system_probe_config.process_service_inference.use_windows_service_name")
```
i opted to rename the variables to `serviceExtractorEnabled`

levin0710 committed:
Passing sysprobeYamlConfig to process check and getting value for useWindowsServiceName

Followed same logic of how the connections check passes down the sysprobeYamlConfig and used it to get config values.

guyarb commented on pkg/process/checks/process.go:
```diff
@@ -147,6 +150,11 @@ func (p *ProcessCheck) Init(syscfg *SysProbeConfig, info *HostInfo, oneShot bool
 
 	p.initConnRates()
 
+	serviceExtractorEnabled := true
```
why is it enabled by default (and hard coded)?

guyarb replied:
Enabling the feature by default and having "process_context" tag for processes can affect USM's service naming.

levin0710 replied:
We are modifying the agent-payload for process. We are including `processContext` into the payload. The logic for extracting `processContext` is in the serviceExtractor, which is used in other checks like the connections checks. For the purposes of the process check, we want to always enable the serviceExtractor because we always want to populate the `processContext` data.

CordeliaLopez replied:
wouldn't we want to enable this only if `process_service_inference` is enabled? 
As far as I know the current behavior only adds `process_context` tags if that feature is explicitly enabled in the configs. This means that if this feature isn't enabled we won't be getting `process_context` and we don't need to populate that field (we don't need serviceExtractorEnabled to be true) since we won't be gathering that data.

levin0710 replied:
As discussed in the slack thread, me and Guy discussed and decided is safe to default true for the agent changes. Will be cautious when integrating this change to the backend.

levin0710 committed:
Merge branch 'main' into levin.sanchez/payload-experimentation

levin0710 committed:
fix dd config error

daniel-taf commented on pkg/process/checks/checks.go:
```diff
@@ -107,7 +107,7 @@ func (p CombinedRunResult) RealtimePayloads() []model.MessageBody {
 func All(config, sysprobeYamlCfg ddconfig.ReaderWriter, syscfg *sysconfigtypes.Config) []Check {
 	return []Check{
-		NewProcessCheck(config),
+		NewProcessCheck(config, sysprobeYamlCfg),
```
Was using the init function not an option for pulling the config values?

levin0710 replied:
I figured it would be more complicated to do so. Also, this is the exact way that the connections check passes this config down, so for consistency purposes this seemed the best.

daniel-taf replied:
Why do you think its more complicated?

levin0710 replied:
I had a harder time following the the call hierarchy of the Init Calls, and it felt like the code change was going to be larger. Also, since this is exactly the way the connections check does it for the same purpose, I didnt see a reason to make the implementation differently.

levin0710 committed:
Merge branch 'main' into levin.sanchez/payload-experimentation

robertjli commented on pkg/process/checks/process.go:
```diff
@@ -40,11 +41,12 @@ const (
+		config:             config,
+		scrubber:           procutil.NewDefaultDataScrubber(),
+		lookupIdProbe:      NewLookupIDProbe(config),
+		sysprobeYamlConfig: sysprobeYamlConfig,
```
Optional: Personally, I'd just create a new service extractor using the config here, and set the `serviceExtractor` field, instead of saving the sysprobe config. It's not used for anything else.

levin0710 replied:
this makes sense to me too

robertjli commented on pkg/process/metadata/parser/service_windows_test.go:
```diff
@@ -58,14 +58,15 @@ func TestWindowsExtractServiceMetadata(t *testing.T) {
 			mockConfig.SetWithoutSource("system_probe_config.process_service_inference.enabled", true)
-
+			enabled := mockConfig.GetBool("system_probe_config.process_service_inference.enabled")
+			useWindowsServiceName := mockConfig.GetBool("system_probe_config.process_service_inference.use_windows_service_name")
```
Please also remove the mock config from this test as it's not necessary

levin0710 replied:
updated!

levin0710 commented on pkg/process/checks/net_test.go:
```diff
@@ -55,7 +55,9 @@ func TestDNSNameEncoding(t *testing.T) {
 	}
-	ex := parser.NewServiceExtractor(ddconfig.MockSystemProbe(t))
+	serviceExtractorEnabled := true
+	useWindowsServiceName := true
```
Yes thay correct, let me change that and verify that still runs as expected

levin0710 committed:
minor fixes

robertjli commented on pkg/process/metadata/parser/service_windows_test.go:
```diff
@@ -56,16 +56,15 @@ func TestWindowsExtractServiceMetadata(t *testing.T) {
-			mockConfig.SetWithoutSource("system_probe_config.process_service_inference.enabled", true)
-
+			enabled := true
+			useWindowsServiceName := true
```
Should be false

levin0710 replied:
enabled should be true, and useWindowsServiceName should be false in this case.

levin0710 committed:
Update service_windows_test.go

levin0710 committed:
Update process.go

levin0710 committed:
format

levin0710 commented:
/merge
